File: be/src/exprs/dictmapping_expr.cpp
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "exprs/dictmapping_expr.h"

namespace starrocks {
DictMappingExpr::DictMappingExpr(const TExprNode& node) : Expr(node, false) {}

StatusOr<ColumnPtr> DictMappingExpr::evaluate_checked(ExprContext* context, Chunk* ptr) {
    // TODO: record rewrite info in ExprContext
    // If dict_func_expr is nullptr, then it means that this DictExpr has not been rewritten.
    // But in some cases we need to evaluate the original expression directly
    // (usually column_expr_predicate).
    if (dict_func_expr == nullptr) {
        return get_child(1)->evaluate_checked(context, ptr);
    } else {
        return dict_func_expr->evaluate_checked(context, ptr);
    }
<<<<<<< HEAD
=======
    // children == 2: DictExpr(string column, string expression) or DictExpr(array column, array column)
    // children == 3: DictExpr(array column, string expression, array expression)
    // do array-expresion first, then string expression
    if (_children.size() == 2) {
        auto target_column = ptr->get_column_by_slot_id(slot_id());
        auto data_column = ColumnHelper::get_data_column(target_column.get());

        if (data_column->is_binary()) {
            DCHECK(dict_func_expr == nullptr);
            return get_child(1)->evaluate_checked(context, ptr);
        } else if (dict_func_expr != nullptr) {
            return dict_func_expr->evaluate_checked(context, ptr);
        } else {
            return Status::InternalError("unreachable path, dict_func_expr shouldn't be nullptr");
        }
    } else if (_children.size() == 3) {
        // array -> string
        ASSIGN_OR_RETURN(auto str, _children[2]->evaluate_checked(context, ptr));
        Chunk cc;
        cc.append_column(str, slot_id());
        if (dict_func_expr != nullptr) {
            return dict_func_expr->evaluate_checked(context, &cc);
        } else {
            return Status::InternalError("unreachable path, array_dict_func_expr shouldn't be nullptr");
        }
    }

    return Status::InternalError(fmt::format("unreachable path, dict children size: {}", _children.size()));
>>>>>>> 4d9cdb7dac ([Enhancement] optimize array_contains/array_position function (#50912))
}

} // namespace starrocks

File: fe/fe-core/src/main/java/com/starrocks/analysis/DictMappingExpr.java
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


package com.starrocks.analysis;

import com.starrocks.common.exception.AnalysisException;
import com.starrocks.thrift.TExprNode;
import com.starrocks.thrift.TExprNodeType;

// DictMappingExpr.
// The original expression will be rewritten as a dictionary mapping function in the global field optimization.
// child(0) was input lowcardinality dictionary column (input was ID type).
// child(1) was origin expr (input was string type).
// 
// in Global Dictionary Optimization. The process of constructing a dictionary mapping requires 
// a new dictionary to be constructed using the global dictionary columns as input columns. 
// So BE needs to know the original expressions.
public class DictMappingExpr extends Expr {

    public DictMappingExpr(Expr ref, Expr call) {
        super(ref);
        this.addChild(ref);
        this.addChild(call);
    }

    public DictMappingExpr(Expr ref, Expr call, Expr stringProvideExpr) {
        super(ref);
        this.addChild(ref);
        this.addChild(call);
        this.addChild(stringProvideExpr);
    }

    protected DictMappingExpr(DictMappingExpr other) {
        super(other);
    }

    @Override
    protected void analyzeImpl(Analyzer analyzer) throws AnalysisException {

    }

    @Override
    protected String toSqlImpl() {
        String fnName = this.type.matchesType(this.getChild(1).getType()) ? "DictDecode" : "DictDefine";

        if (children.size() == 2) {
            return fnName + "(" + this.getChild(0).toSqlImpl() + ", [" + this.getChild(1).toSqlImpl() + "])";
        }
        return fnName + "(" + this.getChild(0).toSqlImpl() + ", [" + this.getChild(1).toSqlImpl() + "], " +
                getChild(2).toSqlImpl() + ")";
    }

    @Override
    protected void toThrift(TExprNode msg) {
        msg.setNode_type(TExprNodeType.DICT_EXPR);
    }

    @Override
    public Expr clone() {
        return new DictMappingExpr(this);
    }
}
